### ISSUE TITLE
Voucher amount not decreased when additional payment is required

### ISSUE THREAD

**Issue by jerome2710** [verbb/gift-voucher#94]
### Describe the bug

We are experiencing issues with one of our major sites when an order is placed with a total value larger than the amount available on the voucher and additional payment is required, the voucher amount is not decreased. The order does not get associated with the voucher.

For example; total order value of $120, voucher amount is $100. Remainder of $20 is paid through the gateway and the order is completed. The remaining voucher amount is still $100.

When the total order value is less or equal, the amount does decrease.

_Note: we have not yet upgraded to Craft 4 due to the size of this project, however we are planning to do so in the near future._

### Steps to reproduce

1. Place an order
2. Apply a voucher code
3. Pay remainder through an available gateway

### Craft CMS version

Craft Pro 3.7.50

### Plugin version

2.7.1

### Multi-site?

Yes

**Comment by engram-design** [comment]
So when a voucher is redeemed, a Redemption is created. Can you check your `giftvoucher_redemptions` against an order you know one was applied to, and let me know if there's any rows for that?

I can't seem to replicate this, no matter what the order vs voucher amount is (less or more). For reference, [here](https://github.com/verbb/gift-voucher/blob/ad83f16827bc055c0664c23cdd0111cd089377da/src/services/CodesService.php#L99) is where we handle our tasks when the order is marked as complete.

What payment gateway are you using? I've seen weird things happen with Mollie and other offsite gateways.

**Comment by jerome2710** [comment]
@engram-design thank you for your reply. There are no rows in the `giftvoucher_redemptions` table for the order that did not decrease the voucher amount. For the other order where the total value was less than the voucher amount and where it was decreased, there is.

The order in Commerce does show the voucher associated, but that might just be the order adjustments snapshot?

![Schermafbeelding 2022-09-02 om 14 27 34](https://user-images.githubusercontent.com/4600095/188142642-de2c479c-8cd5-4b7b-b60b-2738c9f5f2c1.png)

We are in fact using Mollie as a payment gateway, which is one of the larger Western-Europe PSP's.

**Comment by engram-design** [comment]
Very interesting. I've known there are issues with Mollie, but as an Australian resident, I can't get a development account with them which is frustrating! I'm not sure _why_ it's an issue, as it should complete the order in the exact same way another payment gateway does, but something is clearly different. But this could also be a red herring.

> There are no rows in the `giftvoucher_redemptions` table for the order that did not decrease the voucher amount.

So that's what I expect, and points to the [code](https://github.com/verbb/gift-voucher/blob/ad83f16827bc055c0664c23cdd0111cd089377da/src/services/CodesService.php#L99) not running to create that redemption.

Quick question with Mollie - are you redirected away to their site like the old PayPal method? The new PayPal method is through a popup window, but you used to be literally redirected to the PayPal site, then be returned back to the origin site at the end. 

Because we use session-storage to store the vouchers applied to an order ([this PR](https://github.com/verbb/gift-voucher/pull/18)) I'm wondering if a redirection away could clear that session, hence [this check](https://github.com/verbb/gift-voucher/blob/ad83f16827bc055c0664c23cdd0111cd089377da/src/services/CodesService.php#L82) fails. But then it seems odd that there's different behaviour if it's less or greater than the order value.

**Comment by engram-design** [comment]
As something to test, would you be able to modify the code storage behaviour to using a custom field, instead of sessions?

To do this, you'll need to make a custom field (type `Gift Voucher Code`) and add it to your order field layout. This will store the vouchers applied on the order. It should have the handle `giftVoucherCodes`.

Then, add a [gift-voucher.php config file](https://verbb.io/craft-plugins/gift-voucher/docs/get-started/configuration) with the following:

```php
use verbb\giftvoucher\storage\Order;

return [
    'codeStorage' => ['class' => Order::class, 'fieldHandle' => 'giftVoucherCodes'],
];
```

Which switches out session storage for storing against the order. I'd be interested to know if this makes a difference.

**Comment by jerome2710** [comment]
@engram-design I had two weeks leave and just tried your suggestion. I created the field and added it to the order field layout, but it results in the following error:

![Schermafbeelding 2022-09-22 om 17 39 58](https://user-images.githubusercontent.com/4600095/191791920-53853e81-3ec4-4809-8df8-59223d912a08.png)

**Comment by engram-design** [comment]
My mistake @jerome2710 the custom field actually needs to be a `Gift Voucher Code` field, not a plain text one. Sorry for the confusion, I'm not 100% familiar with the functionality of the original PR.

**Comment by jerome2710** [comment]
@engram-design I changed the field type and tried creating some orders. Both fully covered orders as well as orders which required additional payment are now linked to the code, correctly decreasing the remaining amount. The field within the order states _"Nothing selected."_, but I guess this is just a temporary placeholder until the order is completed?

Can we keep this configuration for our site? As you mentioned _"Which switches out session storage for storing against the order."_, it might be useful to mention this site is high traffic, including Varnish, Redis and cached content. The session behavior might be a bit different than a regular simple site.

**Comment by engram-design** [comment]
That's great to know, thanks! And yes, the voucher is removed once the order has been completed. It's essentially a replacement for the session-handling we use to store what's currently applied to a cart. Once a cart has been changed to an order, we no longer need it around.

Yes - you can keep this configuration! And I think even for a high-traffic site, this should be fine. We modify orders when adding the coupon, and when complete.

I think this should address the root cause of the issue with offsite gateways.

**Comment by jerome2710** [comment]
Perfect! I expected the field as just a temporary placeholder, thanks for confirming this.

Thank you for your support, this seems to solve the issue and we will forward the change to our production environment.

**Comment by jerome2710** [comment]
@engram-design, while upgrading to Craft 4 and moving from 2.7.1 to 3.0.0, `\verbb\giftvoucher\models\Settings::$codeStorage` is marked as type string. [Your solution](https://github.com/verbb/gift-voucher/issues/94#issuecomment-1235979536) to use the fieldHandle now triggers an error (Cannot assign array to property of type string). This property isn't mentioned in the documentation either. What would be the correct way to implement this in Craft 4 / 3.0.0?

**Comment by engram-design** [comment]
Fixed for the next release. To get the fix early, change your `verbb/gift-voucher` requirement in `composer.json` to:

```
"require": {
  "verbb/gift-voucher": "dev-craft-4 as 3.0.0",
  "...": "..."
}
```

Then run `composer update`.

**Comment by engram-design** [comment]
Fixed in [3.0.1](https://github.com/verbb/gift-voucher/releases/tag/3.0.1)
